File: clients/python-testcontainers/src/mockly_testcontainers/container.py
import json
import os
import tempfile
import time
import urllib.error
import urllib.parse
import urllib.request
from typing import Any

from testcontainers.core.container import DockerContainer

from ._types import (
    ActiveScenariosResponse,
    CallEntry,
    CallSummary,
    FaultConfig,
    Mock,
    MocklyServerOptions,
    MockRequest,
    MockResponse,
    MockResponsePatch,
    Scenario,
    ScenarioPatch,
)

DEFAULT_IMAGE = "ghcr.io/dever-labs/mockly:latest"
HTTP_PORT = 8090
API_PORT = 9091
CONTAINER_CONFIG_PATH = "/config/mockly.yaml"
DEFAULT_CONFIG = """mockly:
  api:
    port: 9091
protocols:
  http:
    enabled: true
    port: 8090
"""


class MocklyContainer(DockerContainer):
    def __init__(self, image: str = DEFAULT_IMAGE) -> None:
        super().__init__(image)
        self._config_yaml = DEFAULT_CONFIG
        self._config_host_path: str | None = None
        self.with_exposed_ports(HTTP_PORT, API_PORT)

    def with_inline_config(self, yaml: str) -> "MocklyContainer":
        self._config_yaml = yaml
        return self

    def with_options(self, options: MocklyServerOptions) -> "MocklyContainer":
        self._config_yaml = _render_config(HTTP_PORT, API_PORT, options.scenarios)
        return self

    def _configure(self) -> None:
        self._cleanup_config_file()
        with tempfile.NamedTemporaryFile(
            mode="w",
            encoding="utf-8",
            prefix=".mockly-tc-",
            suffix=".yaml",
            dir=os.getcwd(),
            delete=False,
        ) as config_file:
            config_file.write(self._config_yaml)
            self._config_host_path = config_file.name
        self.with_command(f"start -c {CONTAINER_CONFIG_PATH}")
        self.with_volume_mapping(self._config_host_path, CONTAINER_CONFIG_PATH, "ro")

    def _cleanup_config_file(self) -> None:
        if self._config_host_path is None:
            return
        self.volumes.pop(self._config_host_path, None)
        try:
            os.unlink(self._config_host_path)
        except FileNotFoundError:
            pass
        self._config_host_path = None

    def get_http_base(self) -> str:
        host = self.get_container_host_ip()
        port = self.get_exposed_port(HTTP_PORT)
        return f"http://{host}:{port}"

    def get_api_base(self) -> str:
        host = self.get_container_host_ip()
        port = self.get_exposed_port(API_PORT)
        return f"http://{host}:{port}"

    def _wait_ready(self, max_ms: int = 60000) -> None:
        url = f"{self.get_api_base()}/api/protocols"
        deadline = time.monotonic() + max_ms / 1000
        while time.monotonic() < deadline:
            try:
                with urllib.request.urlopen(url, timeout=2):
                    return
            except Exception:
                time.sleep(0.1)
        raise TimeoutError(f"Mockly container did not become ready within {max_ms}ms")

    def start(self) -> "MocklyContainer":
        try:
            self._configure()
            super().start()
            self._wait_ready()
            return self
        except Exception:
            self.stop()
            raise

    def stop(self, force: bool = True, delete_volume: bool = True) -> None:
        try:
            super().stop(force=force, delete_volume=delete_volume)
        finally:
            self._cleanup_config_file()

    def add_mock(self, mock: Mock) -> None:
        self._request("POST", "/api/mocks/http", _mock_to_dict(mock), expected=(200, 201))

    def list_mocks(self) -> list[Mock]:
        return [_parse_mock(item) for item in self._get_json("GET", "/api/mocks/http", expected=(200,))]

    def update_mock(self, mock_id: str, mock: Mock) -> Mock:
        path = f"/api/mocks/http/{urllib.parse.quote(mock_id, safe='')}"
        return _parse_mock(self._get_json("PUT", path, _mock_to_dict(mock), expected=(200,)))

    def patch_mock(self, mock_id: str, patch: MockResponsePatch) -> Mock:
        path = f"/api/mocks/http/{urllib.parse.quote(mock_id, safe='')}"
        return _parse_mock(
            self._get_json(
                "PATCH",
                path,
                _mock_response_patch_to_dict(patch),
                expected=(200,),
            )
        )

    def delete_mock(self, mock_id: str) -> None:
        path = f"/api/mocks/http/{urllib.parse.quote(mock_id, safe='')}"
        self._request("DELETE", path, expected=(200,))

    def get_state(self) -> dict[str, str]:
        return self._get_json("GET", "/api/state", expected=(200,))

    def set_state(self, kv_map: dict[str, str]) -> dict[str, str]:
        return self._get_json("POST", "/api/state", kv_map, expected=(200,))

    def delete_state(self, key: str) -> None:
        path = f"/api/state/{urllib.parse.quote(key, safe='')}"
        self._request("DELETE", path, expected=(200,))

    def reset(self) -> None:
        self._request("POST", "/api/reset", expected=(200,))

    def activate_scenario(self, scenario_id: str) -> None:
        path = f"/api/scenarios/{urllib.parse.quote(scenario_id, safe='')}/activate"
        self._request("POST", path, expected=(200,))

    def deactivate_scenario(self, scenario_id: str) -> None:
        path = f"/api/scenarios/{urllib.parse.quote(scenario_id, safe='')}/deactivate"
        self._request("POST", path, expected=(200,))

    def list_scenarios(self) -> list[Scenario]:
        return [_parse_scenario(item) for item in self._get_json("GET", "/api/scenarios", expected=(200,))]

    def create_scenario(self, scenario: Scenario) -> Scenario:
        return _parse_scenario(
            self._get_json("POST", "/api/scenarios", _scenario_to_dict(scenario), expected=(201,))
        )

    def get_scenario(self, scenario_id: str) -> Scenario:
        path = f"/api/scenarios/{urllib.parse.quote(scenario_id, safe='')}"
        return _parse_scenario(self._get_json("GET", path, expected=(200,)))

    def update_scenario(self, scenario_id: str, scenario: Scenario) -> Scenario:
        path = f"/api/scenarios/{urllib.parse.quote(scenario_id, safe='')}"
        return _parse_scenario(
            self._get_json("PUT", path, _scenario_to_dict(scenario), expected=(200,))
        )

    def delete_scenario(self, scenario_id: str) -> None:
        path = f"/api/scenarios/{urllib.parse.quote(scenario_id, safe='')}"
        self._request("DELETE", path, expected=(200,))

    def list_active_scenarios(self) -> ActiveScenariosResponse:
        data = self._get_json("GET", "/api/scenarios/active", expected=(200,))
        return ActiveScenariosResponse(
            active=list(data.get("active") or []),
            scenarios=[_parse_scenario(item) for item in (data.get("scenarios") or [])],
        )

    def set_fault(self, config: FaultConfig) -> None:
        self._request("POST", "/api/fault/http", _fault_config_to_dict(config), expected=(200,))

    def clear_fault(self) -> None:
        self._request("DELETE", "/api/fault", expected=(200,))

    def get_logs(self, matched_id: str | None = None) -> list[CallEntry]:
        path = _with_optional_matched_id("/api/logs", matched_id)
        return [_parse_call_entry(item) for item in self._get_json("GET", path, expected=(200,))]

    def get_logs_count(self, matched_id: str | None = None) -> int:
        path = _with_optional_matched_id("/api/logs/count", matched_id)
        data = self._get_json("GET", path, expected=(200,))
        return int(data.get("count", 0))

    def clear_logs(self) -> None:
        self._request("DELETE", "/api/logs", expected=(200,))

    def get_calls(self, mock_id: str) -> CallSummary:
        path = f"/api/calls/http/{urllib.parse.quote(mock_id, safe='')}"
        return _parse_call_summary(self._get_json("GET", path, expected=(200,)))

    def clear_calls(self, mock_id: str) -> None:
        path = f"/api/calls/http/{urllib.parse.quote(mock_id, safe='')}"
        self._request("DELETE", path, expected=(200,))

    def clear_all_calls(self) -> None:
        self._request("DELETE", "/api/calls/http", expected=(200,))

    def wait_for_calls(
        self,
        mock_id: str,
        count: int = 1,
        timeout_seconds: int = 10,
    ) -> CallSummary:
        path = f"/api/calls/http/{urllib.parse.quote(mock_id, safe='')}/wait"
        body = {"count": count, "timeout": f"{timeout_seconds}s"}
        return _parse_call_summary(self._get_json("POST", path, body, expected=(200,)))

    def _get_json(
        self,
        method: str,
        path: str,
        body: dict[str, Any] | None = None,
        expected: tuple[int, ...] = (200,),
    ) -> dict[str, Any] | list[Any]:
        if body is None:
            raw = self._request(method, path, expected=expected)
        else:
            raw = self._request(method, path, body, expected=expected)
        return json.loads(raw.decode() if raw else "null")

    def _request(
        self,
        method: str,
        path: str,
        body: dict[str, Any] | None = None,
        expected: tuple[int, ...] = (200,),
    ) -> bytes:
        url = f"{self.get_api_base()}{path}"
        data = json.dumps(body).encode() if body is not None else None
        headers = {"Content-Type": "application/json"} if data is not None else {}
        req = urllib.request.Request(url, data=data, headers=headers, method=method)
        try:
            with urllib.request.urlopen(req) as resp:
                if resp.status not in expected:
                    raise RuntimeError(f"Unexpected status {resp.status} for {method} {path}")
                return resp.read()
        except urllib.error.HTTPError as exc:
            if exc.code in expected:
                return exc.read()
            raise RuntimeError(
                f"HTTP {exc.code} for {method} {path}: {exc.read().decode(errors='replace')}"
            ) from exc


def _render_config(http_port: int, api_port: int, scenarios: list[Scenario] | None) -> str:
    lines = [
        "mockly:",
        "  api:",
        f"    port: {api_port}",
        "protocols:",
        "  http:",
        "    enabled: true",
        f"    port: {http_port}",
    ]

    if scenarios:
        lines.append("scenarios:")
        for scenario in scenarios:
            lines.append(f"  - id: {_yaml_str(scenario.id)}")
            lines.append(f"    name: {_yaml_str(scenario.name)}")
            if scenario.description is not None:
                lines.append(f"    description: {_yaml_str(scenario.description)}")
            if scenario.patches:
                lines.append("    patches:")
                for patch in scenario.patches:
                    lines.append(f"      - mock_id: {_yaml_str(patch.mock_id)}")
                    if patch.status is not None:
                        lines.append(f"        status: {patch.status}")
                    if patch.body is not None:
                        lines.append(f"        body: {_yaml_str(patch.body)}")
                    if patch.headers:
                        lines.append("        headers:")
                        for key, value in patch.headers.items():
                            lines.append(f"          {_yaml_str(key)}: {_yaml_str(value)}")
                    if patch.delay is not None:
                        lines.append(f"        delay: {_yaml_str(patch.delay)}")
                    if patch.disabled is not None:
                        lines.append(f"        disabled: {'true' if patch.disabled else 'false'}")

    return "\n".join(lines) + "\n"


def _yaml_str(value: str) -> str:
    escaped = value.replace("\\", "\\\\").replace('"', '\\"').replace("\n", "\\n")
    return f'"{escaped}"'


def _with_optional_matched_id(path: str, matched_id: str | None) -> str:
    if not matched_id:
        return path
    return f"{path}?matched_id={urllib.parse.quote(matched_id, safe='')}"


def _mock_to_dict(mock: Mock) -> dict[str, Any]:
    body: dict[str, Any] = {
        "id": mock.id,
        "request": {
            "method": mock.request.method,
            "path": mock.request.path,
        },
        "response": _mock_response_to_dict(mock.response),
    }
    if mock.request.headers:
        body["request"]["headers"] = mock.request.headers
    return body


def _mock_response_to_dict(response: MockResponse) -> dict[str, Any]:
    body: dict[str, Any] = {"status": response.status}
    if response.body != "":
        body["body"] = response.body
    if response.headers:
        body["headers"] = response.headers
    if response.delay is not None:
        body["delay"] = response.delay
    return body


def _mock_response_patch_to_dict(patch: MockResponsePatch) -> dict[str, Any]:
    body: dict[str, Any] = {}
    if patch.status is not None:
        body["status"] = patch.status
    if patch.body is not None:
        body["body"] = patch.body
    if patch.headers is not None:
        body["headers"] = patch.headers
    if patch.delay is not None:
        body["delay"] = patch.delay
    return body


def _scenario_to_dict(scenario: Scenario) -> dict[str, Any]:
    body: dict[str, Any] = {
        "id": scenario.id,
        "name": scenario.name,
        "patches": [_scenario_patch_to_dict(patch) for patch in scenario.patches],
    }
    if scenario.description is not None:
        body["description"] = scenario.description
    return body


def _scenario_patch_to_dict(patch: ScenarioPatch) -> dict[str, Any]:
    body: dict[str, Any] = {"mock_id": patch.mock_id}
    if patch.status is not None:
        body["status"] = patch.status
    if patch.body is not None:
        body["body"] = patch.body
    if patch.headers is not None:
        body["headers"] = patch.headers
    if patch.delay is not None:
        body["delay"] = patch.delay
    if patch.disabled is not None:
        body["disabled"] = patch.disabled
    return body


def _fault_config_to_dict(config: FaultConfig) -> dict[str, Any]:
    body: dict[str, Any] = {"enabled": config.enabled}
    if config.delay is not None:
        body["delay"] = config.delay
    if config.status is not None:
        body["status"] = config.status
    if config.error_rate is not None:
        body["error_rate"] = config.error_rate
    return body


def _parse_mock(data: dict[str, Any]) -> Mock:
    request = data.get("request") or {}
    response = data.get("response") or {}
    return Mock(
        id=data.get("id", ""),
        request=MockRequest(
            method=request.get("method", ""),
            path=request.get("path", ""),
            headers=request.get("headers") or {},
        ),
        response=MockResponse(
            status=response.get("status", 0),
            body=response.get("body") or "",
            headers=response.get("headers") or {},
            delay=response.get("delay"),
        ),
    )


def _parse_scenario(data: dict[str, Any]) -> Scenario:
    return Scenario(
        id=data.get("id", ""),
        name=data.get("name", ""),
        description=data.get("description"),
        patches=[_parse_scenario_patch(item) for item in (data.get("patches") or [])],
    )


def _parse_scenario_patch(data: dict[str, Any]) -> ScenarioPatch:
    return ScenarioPatch(
        mock_id=data.get("mock_id", ""),
        status=data.get("status"),
        body=data.get("body"),
        headers=data.get("headers"),
        delay=data.get("delay"),
        disabled=data.get("disabled"),
    )


def _parse_call_summary(data: dict[str, Any]) -> CallSummary:
    return CallSummary(
        mock_id=data.get("mock_id", ""),
        count=data.get("count", 0),
        calls=[_parse_call_entry(item) for item in (data.get("calls") or [])],
    )


def _parse_call_entry(data: dict[str, Any]) -> CallEntry:
    return CallEntry(
        id=data.get("id", ""),
        timestamp=data.get("timestamp", ""),
        protocol=data.get("protocol", ""),
        path=data.get("path", ""),
        duration_ms=data.get("duration_ms", 0),
        method=data.get("method"),
        status=data.get("status"),
        headers=data.get("headers") or {},
        body=data.get("body"),
        matched_id=data.get("matched_id"),
        path_params=data.get("path_params") or {},
    )

File: clients/rust-testcontainers/src/container.rs
use reqwest::blocking::{Client, Response};
use serde::Deserialize;
use serde_json::{Map, Value};
use std::collections::HashMap;
use std::time::Duration;
use testcontainers::core::Container;

use crate::image::{MocklyImage, API_PORT, HTTP_PORT};
use mockly_driver::{
    ActiveScenariosResponse, CallEntry, CallSummary, FaultConfig, Mock, MockResponsePatch, Scenario,
};

pub struct MocklyContainer {
    inner: Container<MocklyImage>,
    client: Client,
}

impl MocklyContainer {
    pub fn new(container: Container<MocklyImage>) -> Self {
        Self {
            inner: container,
            client: Client::new(),
        }
    }

    pub fn inner(&self) -> &Container<MocklyImage> {
        &self.inner
    }

    pub fn into_inner(self) -> Container<MocklyImage> {
        self.inner
    }

    pub fn http_port(&self) -> u16 {
        self.inner
            .get_host_port_ipv4(HTTP_PORT)
            .expect("Mockly HTTP port should be exposed")
    }

    pub fn api_port(&self) -> u16 {
        self.inner
            .get_host_port_ipv4(API_PORT)
            .expect("Mockly API port should be exposed")
    }

    pub fn http_base(&self) -> String {
        format!("http://127.0.0.1:{}", self.http_port())
    }

    pub fn api_base(&self) -> String {
        format!("http://127.0.0.1:{}", self.api_port())
    }

    pub fn add_mock(&self, mock: &Mock) -> Result<(), Box<dyn std::error::Error>> {
        self.expect_status(
            "POST",
            "/api/mocks/http",
            Some(serde_json::to_value(mock)?),
            &[200, 201],
            "add_mock",
        )
    }

    pub fn list_mocks(&self) -> Result<Vec<Mock>, Box<dyn std::error::Error>> {
        self.get_json("/api/mocks/http")
    }

    pub fn update_mock(&self, id: &str, mock: &Mock) -> Result<Mock, Box<dyn std::error::Error>> {
        self.put_and_read(&format!("/api/mocks/http/{id}"), mock)
    }

    pub fn patch_mock(
        &self,
        id: &str,
        patch: &MockResponsePatch,
    ) -> Result<Mock, Box<dyn std::error::Error>> {
        self.patch_and_read(&format!("/api/mocks/http/{id}"), patch)
    }

    pub fn delete_mock(&self, id: &str) -> Result<(), Box<dyn std::error::Error>> {
        self.expect_status(
            "DELETE",
            &format!("/api/mocks/http/{id}"),
            None,
            &[200],
            "delete_mock",
        )
    }

    pub fn get_state(&self) -> Result<HashMap<String, String>, Box<dyn std::error::Error>> {
        self.get_json("/api/state")
    }

    pub fn set_state(
        &self,
        state: &HashMap<String, String>,
    ) -> Result<HashMap<String, String>, Box<dyn std::error::Error>> {
        self.post_and_read("/api/state", state)
    }

    pub fn delete_state(&self, key: &str) -> Result<(), Box<dyn std::error::Error>> {
        self.expect_status(
            "DELETE",
            &format!("/api/state/{key}"),
            None,
            &[200],
            "delete_state",
        )
    }

    pub fn reset(&self) -> Result<(), Box<dyn std::error::Error>> {
        self.expect_status("POST", "/api/reset", None, &[200], "reset")
    }

    pub fn activate_scenario(&self, id: &str) -> Result<(), Box<dyn std::error::Error>> {
        self.expect_status(
            "POST",
            &format!("/api/scenarios/{id}/activate"),
            None,
            &[200],
            "activate_scenario",
        )
    }

    pub fn deactivate_scenario(&self, id: &str) -> Result<(), Box<dyn std::error::Error>> {
        self.expect_status(
            "POST",
            &format!("/api/scenarios/{id}/deactivate"),
            None,
            &[200],
            "deactivate_scenario",
        )
    }

    pub fn list_scenarios(&self) -> Result<Vec<Scenario>, Box<dyn std::error::Error>> {
        self.get_json("/api/scenarios")
    }

    pub fn create_scenario(
        &self,
        scenario: &Scenario,
    ) -> Result<Scenario, Box<dyn std::error::Error>> {
        self.post_and_read_with_status("/api/scenarios", scenario, 201)
    }

    pub fn get_scenario(&self, id: &str) -> Result<Scenario, Box<dyn std::error::Error>> {
        self.get_json(&format!("/api/scenarios/{id}"))
    }

    pub fn update_scenario(
        &self,
        id: &str,
        scenario: &Scenario,
    ) -> Result<Scenario, Box<dyn std::error::Error>> {
        self.put_and_read(&format!("/api/scenarios/{id}"), scenario)
    }

    pub fn delete_scenario(&self, id: &str) -> Result<(), Box<dyn std::error::Error>> {
        self.expect_status(
            "DELETE",
            &format!("/api/scenarios/{id}"),
            None,
            &[200],
            "delete_scenario",
        )
    }

    pub fn list_active_scenarios(
        &self,
    ) -> Result<ActiveScenariosResponse, Box<dyn std::error::Error>> {
        self.get_json("/api/scenarios/active")
    }

    pub fn set_fault(&self, config: &FaultConfig) -> Result<(), Box<dyn std::error::Error>> {
        self.expect_status(
            "POST",
            "/api/fault/http",
            Some(fault_payload(config)),
            &[200],
            "set_fault",
        )
    }

    pub fn clear_fault(&self) -> Result<(), Box<dyn std::error::Error>> {
        self.expect_status("DELETE", "/api/fault", None, &[200, 204], "clear_fault")
    }

    pub fn get_logs(
        &self,
        matched_id: Option<&str>,
    ) -> Result<Vec<CallEntry>, Box<dyn std::error::Error>> {
        let url = self.with_optional_matched_id("/api/logs", matched_id)?;
        self.get_json_from_url(url, "/api/logs")
    }

    pub fn clear_logs(&self) -> Result<(), Box<dyn std::error::Error>> {
        self.expect_status("DELETE", "/api/logs", None, &[200], "clear_logs")
    }

    pub fn get_logs_count(
        &self,
        matched_id: Option<&str>,
    ) -> Result<i64, Box<dyn std::error::Error>> {
        let url = self.with_optional_matched_id("/api/logs/count", matched_id)?;
        let response: CountResponse = self.get_json_from_url(url, "/api/logs/count")?;
        Ok(response.count)
    }

    pub fn get_calls(&self, mock_id: &str) -> Result<CallSummary, Box<dyn std::error::Error>> {
        self.get_json(&format!("/api/calls/http/{mock_id}"))
    }

    pub fn clear_calls(&self, mock_id: &str) -> Result<(), Box<dyn std::error::Error>> {
        self.expect_status(
            "DELETE",
            &format!("/api/calls/http/{mock_id}"),
            None,
            &[200],
            "clear_calls",
        )
    }

    pub fn clear_all_calls(&self) -> Result<(), Box<dyn std::error::Error>> {
        self.expect_status("DELETE", "/api/calls/http", None, &[200], "clear_all_calls")
    }

    pub fn wait_for_calls(
        &self,
        mock_id: &str,
        count: usize,
        timeout: Duration,
    ) -> Result<CallSummary, Box<dyn std::error::Error>> {
        let resp = self.request(
            "POST",
            &format!("/api/calls/http/{mock_id}/wait"),
            Some(serde_json::json!({
                "count": count,
                "timeout": format!("{}ms", timeout.as_millis()),
            })),
        )?;
        let status = resp.status().as_u16();
        let body = resp.text()?;
        if status == 408 {
            return Err(format!(
                "wait_for_calls: timeout waiting for {count} call(s) on '{mock_id}'"
            )
            .into());
        }
        if status != 200 {
            return Err(
                format!("POST /api/calls/http/{mock_id}/wait failed: {status}: {body}").into(),
            );
        }
        Ok(serde_json::from_str(&body)?)
    }

    fn get_json<T: serde::de::DeserializeOwned>(
        &self,
        path: &str,
    ) -> Result<T, Box<dyn std::error::Error>> {
        let resp = self.request("GET", path, None)?;
        let status = resp.status().as_u16();
        let body = resp.text()?;
        if status == 200 {
            Ok(serde_json::from_str(&body)?)
        } else {
            Err(format!("GET {path} failed: {status}: {body}").into())
        }
    }

    fn get_json_from_url<T: serde::de::DeserializeOwned>(
        &self,
        url: reqwest::Url,
        path: &str,
    ) -> Result<T, Box<dyn std::error::Error>> {
        let resp = self.request_url("GET", url, None)?;
        let status = resp.status().as_u16();
        let body = resp.text()?;
        if status == 200 {
            Ok(serde_json::from_str(&body)?)
        } else {
            Err(format!("GET {path} failed: {status}: {body}").into())
        }
    }

    fn post_and_read<T: serde::de::DeserializeOwned>(
        &self,
        path: &str,
        body: &impl serde::Serialize,
    ) -> Result<T, Box<dyn std::error::Error>> {
        self.post_and_read_with_status(path, body, 200)
    }

    fn post_and_read_with_status<T: serde::de::DeserializeOwned>(
        &self,
        path: &str,
        body: &impl serde::Serialize,
        expected_status: u16,
    ) -> Result<T, Box<dyn std::error::Error>> {
        let resp = self.request("POST", path, Some(serde_json::to_value(body)?))?;
        self.read_json_response(resp, "POST", path, expected_status)
    }

    fn put_and_read<T: serde::de::DeserializeOwned>(
        &self,
        path: &str,
        body: &impl serde::Serialize,
    ) -> Result<T, Box<dyn std::error::Error>> {
        let resp = self.request("PUT", path, Some(serde_json::to_value(body)?))?;
        self.read_json_response(resp, "PUT", path, 200)
    }

    fn patch_and_read<T: serde::de::DeserializeOwned>(
        &self,
        path: &str,
        body: &impl serde::Serialize,
    ) -> Result<T, Box<dyn std::error::Error>> {
        let resp = self.request("PATCH", path, Some(serde_json::to_value(body)?))?;
        self.read_json_response(resp, "PATCH", path, 200)
    }

    fn read_json_response<T: serde::de::DeserializeOwned>(
        &self,
        resp: Response,
        method: &str,
        path: &str,
        expected_status: u16,
    ) -> Result<T, Box<dyn std::error::Error>> {
        let status = resp.status().as_u16();
        let body = resp.text()?;
        if status == expected_status {
            Ok(serde_json::from_str(&body)?)
        } else {
            Err(format!("{method} {path} failed: {status}: {body}").into())
        }
    }

    fn with_optional_matched_id(
        &self,
        path: &str,
        matched_id: Option<&str>,
    ) -> Result<reqwest::Url, Box<dyn std::error::Error>> {
        let mut url = reqwest::Url::parse(&format!("{}{}", self.api_base(), path))?;
        if let Some(matched_id) = matched_id {
            url.query_pairs_mut().append_pair("matched_id", matched_id);
        }
        Ok(url)
    }

    fn expect_status(
        &self,
        method: &str,
        path: &str,
        body: Option<Value>,
        expected: &[u16],
        op: &str,
    ) -> Result<(), Box<dyn std::error::Error>> {
        let resp = self.request(method, path, body)?;
        let status = resp.status().as_u16();
        if expected.contains(&status) {
            Ok(())
        } else {
            let body = resp.text().unwrap_or_default();
            Err(format!(
                "{op} failed: expected one of {:?}, got {status}: {body}",
                expected
            )
            .into())
        }
    }

    fn request(
        &self,
        method: &str,
        path: &str,
        body: Option<Value>,
    ) -> Result<Response, Box<dyn std::error::Error>> {
        let url = reqwest::Url::parse(&format!("{}{}", self.api_base(), path))?;
        self.request_url(method, url, body)
    }

    fn request_url(
        &self,
        method: &str,
        url: reqwest::Url,
        body: Option<Value>,
    ) -> Result<Response, Box<dyn std::error::Error>> {
        let builder = match method {
            "GET" => self.client.get(url),
            "POST" => self.client.post(url),
            "PUT" => self.client.put(url),
            "PATCH" => self.client.patch(url),
            "DELETE" => self.client.delete(url),
            other => return Err(format!("unsupported HTTP method: {other}").into()),
        };

        let builder = match body {
            Some(body) => builder.json(&body),
            None => builder,
        };

        Ok(builder.send()?)
    }
}

impl From<Container<MocklyImage>> for MocklyContainer {
    fn from(container: Container<MocklyImage>) -> Self {
        Self::new(container)
    }
}

fn fault_payload(config: &FaultConfig) -> Value {
    let mut map = Map::new();

    if let Some(delay) = &config.delay {
        map.insert("delay".to_string(), Value::String(delay.clone()));
    }
    if let Some(status) = config.status {
        map.insert("status".to_string(), Value::Number(status.into()));
    }
    if let Some(error_rate) = config.error_rate {
        if let Some(number) = serde_json::Number::from_f64(error_rate) {
            map.insert("error_rate".to_string(), Value::Number(number));
        }
    }

    Value::Object(map)
}

#[derive(Deserialize)]
struct CountResponse {
    count: i64,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fault_payload_maps_status_override_to_status() {
        let payload = fault_payload(&FaultConfig {
            enabled: true,
            delay: Some("25ms".to_string()),
            status: Some(503),
            error_rate: Some(0.25),
        });

        assert_eq!(payload["delay"], "25ms");
        assert_eq!(payload["status"], 503);
        assert_eq!(payload["error_rate"], 0.25);
        assert!(payload.get("status_override").is_none());
        assert!(payload.get("enabled").is_none());
    }
}
